libs/SalesforceSDKCore/SalesforceSDKCore/Classes/URLHandlers/SFSDKAdvancedAuthURLHandler.m

```diff
@@ -0,0 +1,58 @@
+/*
+ SFSDKAdvancedAuthURLHandler.m
+ SalesforceSDKCore
+ 
+ Created by Raj Rao on 8/28/17.
+ 
+ Copyright (c) 2017-present, salesforce.com, inc. All rights reserved.
+ 
+ Redistribution and use of this software in source and binary forms, with or without modification,
+ are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this list of conditions
+ and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice, this list of
+ conditions and the following disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#import "SFSDKURLHandler.h"
+#import "SFSDKAdvancedAuthURLHandler.h"
+#import "SFSDKOAuthClient.h"
+#import "SFSDKAuthPreferences.h"
+#import "SFUserAccountManager+Internal.h"
+#import "NSURL+SFAdditions.h"
+#import "SFSDKOAuthClientConfig.h"
+
+@implementation SFSDKAdvancedAuthURLHandler
+
+- (BOOL)canHandleRequest:(NSURL *)url options:(NSDictionary *)options {
```

Comment:
Question about all our URLs:
do they look like`app://?arg1=val1&arg2=val2`
Should they instead look like: `app://action?arg1=val1&arg2=val2` ?
If there was an action there, we could use that in the SFSDKURLHandler classes to decide the URL can be processed or not.

Reply:
@wmathurin Our urls are all based of ```CONNECTEDAPP_REDIRECT_URL?arg1=val1&arg2..```.  I'd prefer to either construct the entire URL path or leave it intact as provided.  Since we accept the callback url from developer, the preferred method would be to tinker with query params.  I did initially consider adding a CMD parameter to the query strings but didn't see much upfront value to it since there were only 3-4 actions and in the case of Advanced Auth (legacy) this param won't exist.  I still think its do-able for the other 3 actions (IDP req, IDP response & error).

Reply:
Right now it is as if we had function signatures without function names.
We should have a command name in there either as another query param or in the pass.
The URL handler could simply look for the name to know whether they can handle the URL or not (without having to know anything about the rest of the URL). Or if we want them to really "handle" the URL, they could be responsible for parsing the query params out of the URL and calling methods that take those params directly instead of the URL.

Are we expecting to get just the custom scheme of the idp (or sp) app in all the URL building methods or a scheme with a path (so xxx or xxx://yyy)?

We should think about versioning - what if more parameters are needed in the future, if we have command name, we could simply come up with a new command name for the new signature otherwise, we will have to some logic checking the presence or absence of certain query params.